In [1]:
import pandas as pd



A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.0.2 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "/home/codespace/anaconda3/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/codespace/anaconda3/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/codespace/anaconda3/lib/python3.9/site-packages/ipykernel_launcher.py", line 16, in <module>
    app.launch_new_instance()
  File "/home/codespace/anaconda3/lib/python3.9/site-packages/traitlets/config/application.py", line 846, in launch_instance
    app.start()
  

AttributeError: _ARRAY_API not found


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.0.2 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "/home/codespace/anaconda3/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/codespace/anaconda3/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/codespace/anaconda3/lib/python3.9/site-packages/ipykernel_launcher.py", line 16, in <module>
    app.launch_new_instance()
  File "/home/codespace/anaconda3/lib/python3.9/site-packages/traitlets/config/application.py", line 846, in launch_instance
    app.start()
  

AttributeError: _ARRAY_API not found

In [2]:
from sklearn.feature_extraction import DictVectorizer
from sklearn.linear_model import LinearRegression
from sklearn.metrics import root_mean_squared_error


In [3]:
pd.__version__

'2.3.0+4.g1dfc98e16a'

### Download datasets

In [1]:
!wget https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-02.parquet

--2025-06-13 00:46:00--  https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-02.parquet
Resolving d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)... 18.239.38.163, 18.239.38.83, 18.239.38.181, ...
Connecting to d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)|18.239.38.163|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 47748012 (46M) [application/x-www-form-urlencoded]
Saving to: ‘yellow_tripdata_2023-02.parquet’

yellow_tripdata_202 100%[===================>]  45.54M  27.5MB/s    in 1.7s    

2025-06-13 00:46:02 (27.5 MB/s) - ‘yellow_tripdata_2023-02.parquet’ saved [47748012/47748012]



In [3]:
!wget https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-01.parquet

--2025-06-13 00:48:46--  https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-01.parquet
Resolving d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)... 18.239.38.147, 18.239.38.83, 18.239.38.181, ...
Connecting to d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)|18.239.38.147|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 47673370 (45M) [application/x-www-form-urlencoded]
Saving to: ‘yellow_tripdata_2023-01.parquet’

yellow_tripdata_202 100%[===================>]  45.46M   256MB/s    in 0.2s    

2025-06-13 00:48:46 (256 MB/s) - ‘yellow_tripdata_2023-01.parquet’ saved [47673370/47673370]



## Read Data
   

In [4]:
df_jan_2023 = pd.read_parquet('./yellow_tripdata_2023-01.parquet')


In [5]:
df_jan_2023

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1               2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2               2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3               1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4               2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   
...           ...                  ...                   ...              ...   
3066761         2  2023-01-31 23:58:34   2023-02-01 00:12:33              NaN   
3066762         2  2023-01-31 23:31:09   2023-01-31 23:50:36              NaN   
3066763         2  2023-01-31 23:01:05   2023-01-31 23:25:36              NaN   
3066764         2  2023-01-31 23:40:00   2023-01-31 23:53:00              NaN   
3066765         2  2023-01-31 23:07:32   2023-01-31 23:21:56              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 0.97         1.0                  N           161   
1                 1.10         1.0                  N            43   
2                 2.51         1.0                  N            48   
3                 1.90         1.0                  N           138   
4                 1.43         1.0                  N           107   
...                ...         ...                ...           ...   
3066761           3.05         NaN               None           107   
3066762           5.80         NaN               None           112   
3066763           4.67         NaN               None           114   
3066764           3.15         NaN               None           230   
3066765           2.85         NaN               None           262   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 141             2         9.30   1.00      0.5        0.00   
1                 237             1         7.90   1.00      0.5        4.00   
2                 238             1        14.90   1.00      0.5       15.00   
3                   7             1        12.10   7.25      0.5        0.00   
4                  79             1        11.40   1.00      0.5        3.28   
...               ...           ...          ...    ...      ...         ...   
3066761            48             0        15.80   0.00      0.5        3.96   
3066762            75             0        22.43   0.00      0.5        2.64   
3066763           239             0        17.61   0.00      0.5        5.32   
3066764            79             0        18.15   0.00      0.5        4.43   
3066765           143             0        15.97   0.00      0.5        2.00   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    1.0         14.30   
1                 0.0                    1.0         16.90   
2                 0.0                    1.0         34.90   
3                 0.0                    1.0         20.85   
4                 0.0                    1.0         19.68   
...               ...                    ...           ...   
3066761           0.0                    1.0         23.76   
3066762           0.0                    1.0         29.07   
3066763           0.0                    1.0         26.93   
3066764           0.0                    1.0         26.58   
3066765           0.0                    1.0         21.97   

         congestion_surcharge  airport_fee  
0                         2.5         0.00  
1                         2.5         0.00  
2                         2.5         0.00  
3                         0.0         1.25  
4                         2.5         0.00  
...                       ...          ...  
3066761                   NaN          NaN  
3066762                   NaN          NaN  
3066763                   NaN          NaN  
3066764                   N

### 1) Read the data for January. How many columns are there?

In [6]:
len(df_jan_2023.columns)

19

In [7]:
len(df_jan_2023)

3066766

In [8]:
df_jan_2023['duration'] = df_jan_2023.tpep_dropoff_datetime - df_jan_2023.tpep_pickup_datetime


In [9]:
df_jan_2023['duration']

0         0 days 00:08:26
1         0 days 00:06:19
2         0 days 00:12:45
3         0 days 00:09:37
4         0 days 00:10:50
                ...      
3066761   0 days 00:13:59
3066762   0 days 00:19:27
3066763   0 days 00:24:31
3066764   0 days 00:13:00
3066765   0 days 00:14:24
Name: duration, Length: 3066766, dtype: timedelta64[us]

### 2) What's the standard deviation of the trips duration in January?

In [10]:
df_jan_2023['duration'].describe()

count                   3066766
mean     0 days 00:15:40.139710
std      0 days 00:42:35.661074
min           -1 days +23:30:48
25%             0 days 00:07:07
50%             0 days 00:11:31
75%             0 days 00:18:18
max             6 days 23:09:11
Name: duration, dtype: object

std is 42:35

### 3) What fraction of the records left after you dropped the outliers?

In [11]:
df_jan_2023.duration = df_jan_2023.duration.apply(lambda td: td.total_seconds() / 60)

In [12]:
df_jan_2023 = df_jan_2023[(df_jan_2023.duration >= 1) & (df_jan_2023.duration <= 60)]

In [13]:
len(df_jan_2023)

3009173

In [14]:
3009173 / 3066766

0.9812202822125979

#### The fraction is 98%

### 4) What's the dimensionality of this matrix (number of columns)?

In [15]:
categorical = ['PULocationID', 'DOLocationID']

df_jan_2023[categorical] = df_jan_2023[categorical].astype(str)

/tmp/ipykernel_20256/905221912.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_jan_2023[categorical] = df_jan_2023[categorical].astype(str)


In [16]:
train_dicts = df_jan_2023[categorical].to_dict(orient='records')

In [17]:
dv = DictVectorizer()


In [18]:
X_train = dv.fit_transform(train_dicts)

In [19]:
X_train.shape

(3009173, 515)

#### Number of columns is 515

### 5) What's the RMSE on train?


In [20]:
target = 'duration'
y_train = df_jan_2023[target].values

lr = LinearRegression()
lr.fit(X_train, y_train)

y_pred = lr.predict(X_train)

root_mean_squared_error(y_train, y_pred)

7.6492624397080675

#### The RMSE is 7.64

### 6) What's the RMSE on validation?

#### Load Validation DataSet

In [21]:
df_feb_2023 = pd.read_parquet('./yellow_tripdata_2023-02.parquet')
df_feb_2023

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2023-02-01 00:32:53   2023-02-01 00:34:34              2.0   
1               2  2023-02-01 00:35:16   2023-02-01 00:35:30              1.0   
2               2  2023-02-01 00:35:16   2023-02-01 00:35:30              1.0   
3               1  2023-02-01 00:29:33   2023-02-01 01:01:38              0.0   
4               2  2023-02-01 00:12:28   2023-02-01 00:25:46              1.0   
...           ...                  ...                   ...              ...   
2913950         2  2023-02-28 23:46:00   2023-03-01 00:05:00              NaN   
2913951         2  2023-02-28 23:26:02   2023-02-28 23:37:10              NaN   
2913952         2  2023-02-28 23:24:00   2023-02-28 23:38:00              NaN   
2913953         2  2023-02-28 23:03:00   2023-02-28 23:10:00              NaN   
2913954         2  2023-02-28 23:03:03   2023-02-28 23:12:51              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 0.30         1.0                  N           142   
1                 0.00         1.0                  N            71   
2                 0.00         1.0                  N            71   
3                18.80         1.0                  N           132   
4                 3.22         1.0                  N           161   
...                ...         ...                ...           ...   
2913950           4.65         NaN               None           249   
2913951           2.47         NaN               None           186   
2913952           3.49         NaN               None           158   
2913953           2.13         NaN               None            79   
2913954           2.28         NaN               None           161   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 163             2         4.40   3.50      0.5        0.00   
1                  71             4        -3.00  -1.00     -0.5        0.00   
2                  71             4         3.00   1.00      0.5        0.00   
3                  26             1        70.90   2.25      0.5        0.00   
4                 145             1        17.00   1.00      0.5        3.30   
...               ...           ...          ...    ...      ...         ...   
2913950           140             0        20.22   0.00      0.5        4.84   
2913951            79             0        13.66   0.00      0.5        2.65   
2913952           143             0        17.64   0.00      0.5        0.00   
2913953           162             0        13.56   0.00      0.5        2.63   
2913954           140             0        14.89   0.00      0.5        3.78   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    1.0          9.40   
1                 0.0                   -1.0         -5.50   
2                 0.0                    1.0          5.50   
3                 0.0                    1.0         74.65   
4                 0.0                    1.0         25.30   
...               ...                    ...           ...   
2913950           0.0                    1.0         29.06   
2913951           0.0                    1.0         20.31   
2913952           0.0                    1.0         21.64   
2913953           0.0                    1.0         20.19   
2913954           0.0                    1.0         22.67   

         congestion_surcharge  Airport_fee  
0                         2.5         0.00  
1                         0.0         0.00  
2                         0.0         0.00  
3                         0.0         1.25  
4                         2.5         0.00  
...                       ...          ...  
2913950                   NaN          NaN  
2913951                   NaN          NaN  
2913952                   NaN          NaN  
2913953                   N

In [22]:
df_feb_2023['duration'] = df_feb_2023.tpep_dropoff_datetime - df_feb_2023.tpep_pickup_datetime

df_feb_2023['duration']

0         0 days 00:01:41
1         0 days 00:00:14
2         0 days 00:00:14
3         0 days 00:32:05
4         0 days 00:13:18
                ...      
2913950   0 days 00:19:00
2913951   0 days 00:11:08
2913952   0 days 00:14:00
2913953   0 days 00:07:00
2913954   0 days 00:09:48
Name: duration, Length: 2913955, dtype: timedelta64[us]

In [23]:
df_feb_2023.duration = df_feb_2023.duration.apply(lambda td: td.total_seconds() / 60)
df_feb_2023 = df_feb_2023[(df_feb_2023.duration >= 1) & (df_feb_2023.duration <= 60)]

In [24]:
categorical = ['PULocationID', 'DOLocationID']

df_feb_2023[categorical] = df_feb_2023[categorical].astype(str)

/tmp/ipykernel_20256/2305748683.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_feb_2023[categorical] = df_feb_2023[categorical].astype(str)


### Note: 
Use dv.transform not dv.fittransform because you want the vectorizer to return the same dimentions for the validation dataset

In [25]:
val_dicts = df_feb_2023[categorical].to_dict(orient='records')
X_val = dv.transform(val_dicts)
X_val.shape

(2855951, 515)

In [27]:
target = 'duration'
y_val = df_feb_2023[target].values

y_pred = lr.predict(X_val)

root_mean_squared_error(y_val, y_pred)

7.81181211389241

### The RMSE on validation is 7.81